src/Stratis.Bitcoin/Base/BackgroundWork/BackgroundWorkProvider.cs

```diff
@@ -0,0 +1,214 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Signals;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Base.BackgroundWork
+{
+    /// <summary>
+    /// Provides functionality for creating and tracking operations that happens in background.
+    /// </summary>
+    public partial class BackgroundWorkProvider : IBackgroundWorkProvider
+    {
+        private const int DefaultLoopRepeatInterval = 1000;
+
+        private object lockAsyncDelegates;
+        private object lockAsyncLoops;
+
+        /// <summary>
+        /// Holds a list of currently running async delegates or delegates that stopped because of unhandled exceptions.
+        /// Protected by <see cref="lockAsyncDelegates"/> lock
+        /// </summary>
+        private Dictionary<IAsyncDelegateWorker, AsyncTaskInfo> asyncDelegateWorkers;
+
+        /// <summary>
+        /// Holds a list of currently running async loops or loops that stopped because of unhandled exceptions.
+        /// Protected by <see cref="lockAsyncLoops"/> lock
+        /// </summary>
+        private Dictionary<IAsyncLoop, AsyncTaskInfo> asyncLoops;
+
+        private ILoggerFactory loggerFactory;
+        private ILogger logger;
+        private ISignals signals;
+        private readonly INodeLifetime nodeLifetime;
+        private readonly IAsyncLoopFactory asyncLoopFactory;
+
+        public BackgroundWorkProvider(ILoggerFactory loggerFactory, ISignals signals, INodeLifetime nodeLifetime, IAsyncLoopFactory asyncLoopFactory)
+        {
+            this.lockAsyncDelegates = new object();
+            this.lockAsyncLoops = new object();
+
+            this.asyncDelegateWorkers = new Dictionary<IAsyncDelegateWorker, AsyncTaskInfo>();
+
+            this.loggerFactory = Guard.NotNull(loggerFactory, nameof(loggerFactory));
+            this.logger = this.loggerFactory.CreateLogger(nameof(BackgroundWorkProvider));
+
+            this.signals = Guard.NotNull(signals, nameof(signals));
+            this.nodeLifetime = Guard.NotNull(nodeLifetime, nameof(nodeLifetime));
+            this.asyncLoopFactory = Guard.NotNull(asyncLoopFactory, nameof(asyncLoopFactory));
+        }
+
+        /// <inheritdoc />
+        public IAsyncDelegateWorker CreateAndRunAsyncDelegate<T>(string friendlyName, Func<T, CancellationToken, Task> @delegate)
+        {
+            AsyncQueue<T> newDelegate;
+
+            lock (this.lockAsyncDelegates)
+            {
+                newDelegate = new AsyncQueue<T>(new AsyncQueue<T>.OnEnqueueAsync(@delegate));
```

Comment:
The first 3 lines dont have to be inside the lock (not a big deal tough)

Reply:
actually it must be added to the dictionary before I set the task continuation, because task start running in AsynQueue constructor, so I had changed it anyway and put 2 lines out of lock (not yet committed)